packages/block-editor/src/store/selectors.js

```diff
@@ -1565,14 +1565,25 @@ const canInsertBlockTypeUnmemoized = (
 		return false;
 	}
 
-	const parentAllowedBlocks = parentBlockListSettings?.allowedBlocks;
-	const hasParentAllowedBlock = checkAllowList(
-		parentAllowedBlocks,
+	const parentName = getBlockName( state, rootClientId );
+	const parentBlockType = getBlockType( parentName );
+	const parentAllowedChildBlocks = parentBlockType?.children;
+	let hasParentAllowedBlock = checkAllowList(
+		parentAllowedChildBlocks,
 		blockName
 	);
 
+	// If decision can't be made with `blockType.children` (i.e., it's not there)
+	// decide using the `allowedBlocks` block list setting (as a legacy fallback).
+	if ( hasParentAllowedBlock === null ) {
+		const parentAllowedBlocks = parentBlockListSettings?.allowedBlocks;
+		hasParentAllowedBlock = checkAllowList(
+			parentAllowedBlocks,
+			blockName
+		);
+	}
+
```

Comment:
I appreciate you've added this fallback (and the rationale for the name change), yet I still wonder if the value of the name change outweighs the additional code that would need to remain indefinitely to cover the backward compatibility. There's also the long tail of documentation out there that is referencing `allowedBlocks`.

Reply:
I think the name change is not related to this. Until now the `allowedBlocks` were a part of the block's `edit` UI, passed as a prop to the `<InnerBlocks>` component. Example:
```js
function ColumnsEdit() {
  const blockProps = useBlockProps();
  const props = useInnerBlocksProps( blockProps, { allowedBlocks: [ 'core/column' ] } );
  return <div { ...props } />;
}
```
This way the `allowedBlocks` info is buried inside React code and is not extensible.

The core idea of this PR is to move this info to `block.json`. Currently I'm adding it as a `children` field, but even if it was `allowedBlocks`, it remains true that we must continue to support `InnerBlocks.allowedBlocks`, and that there is a long tail of documentation that references `InnerBlocks.allowedBlocks`.

Reply:
How confusing would it be if we rename `children` to `allowedBlocks` and lean on the idea proposed in that PR that if `allowedBlocks` is not defined in block.json, the prop passed to `InnerBlocks` will be used, but if it's defined, it would take precedence?

Reply:
I'd support using `allowedBlocks` instead of `children`. It's immediately clear what it means, and I don't think it's confusing at all, given that we already have the concept of `allowedBlocks`. We're just moving the place where they're defined. They even have the same shape!